jerry-core/ecma/operations/ecma-typedarray-object.c

```diff
@@ -0,0 +1,882 @@
+
+      ecma_value_t arg3 = ((arguments_list_len > 2) ? arguments_list_p[2]
+                                                    : ecma_make_simple_value (ECMA_SIMPLE_VALUE_UNDEFINED));
+
+      ECMA_OP_TO_NUMBER_TRY_CATCH (num2, arg2, ret);
+
+      uint32_t offset = ecma_number_to_uint32 (num2);
+
+      if (ecma_number_is_negative (ecma_number_to_int32 (num2)) || (offset % bytes_in_element != 0))
+      {
+        return ecma_raise_range_error (ECMA_ERR_MSG ("Invalid offset."));
+      }
+
+      ecma_length_t buf_byte_length = ecma_arraybuffer_get_length (arraybuffer_p);
+      ecma_length_t new_byte_length;
+
+      if (ecma_is_value_undefined (arg3))
+      {
+        if (buf_byte_length % bytes_in_element != 0)
+        {
+          return ecma_raise_range_error (ECMA_ERR_MSG ("Invalid length."));
+        }
+
+        if (buf_byte_length < offset)
+        {
+          return ecma_raise_range_error (ECMA_ERR_MSG ("Invalid length."));
+        }
+
+        new_byte_length = (ecma_length_t) (buf_byte_length - offset);
+      }
+      else
+      {
+        ECMA_OP_TO_NUMBER_TRY_CATCH (num3, arg3, ret);
+        int32_t new_length = ecma_number_to_int32 (num3);
+        new_length = (new_length > 0) ? new_length : 0;
+        new_byte_length = (ecma_length_t) new_length * bytes_in_element;
+
+        if (new_byte_length + offset > buf_byte_length)
+        {
+          return ecma_raise_range_error (ECMA_ERR_MSG ("Invalid length."));
+        }
+
+        ECMA_OP_TO_NUMBER_FINALIZE (num3);
+
+        if (!ecma_is_value_empty (ret))
+        {
+          return ret;
+        }
+      }
+
+      ret = ecma_make_object_value (ecma_typedarray_create_object_with_buffer (arraybuffer_p,
+                                                                               offset,
+                                                                               new_byte_length / bytes_in_element,
+                                                                               proto_p,
+                                                                               bytes_in_element,
+                                                                               class_id));
+
+      ECMA_OP_TO_NUMBER_FINALIZE (num2);
+    }
+    else
+    {
+      /* 22.2.1.4 */
+      ecma_value_t undef_val = ecma_make_simple_value (ECMA_SIMPLE_VALUE_UNDEFINED);
+      ret = ecma_op_typedarray_from (arguments_list_p[0],
+                                     undef_val,
+                                     undef_val,
+                                     proto_p,
+                                     bytes_in_element,
+                                     class_id);
+    }
+  }
+
+  return ret;
+} /* ecma_op_create_typedarray */
+
+/**
+ * Check if the value is typedarray
+ *
+ * @return True: it is a typedarray object
+ *         False: it is not object or it is not a typedarray
+ */
+bool
+ecma_is_typedarray (ecma_value_t target) /**< the target need to be checked */
```

Comment:
yes,  and I think just use one byte can do the all things for typedarray for now

first 4 bits indicate the class group of the object (e.g. typedarray) and it can support 16 different group for future extension

1 bit indicate whether the object need allocate other space for length and offset

3 bits (0-7) show the item-length (1 - 64)